---
title: "R Notebook"
output: html_notebook
---


# Setup


## Libraries 

```{r}
library(tidyverse)
library(SpatialFeatureExperiment)
library(Voyager)
library(alabaster.sfe)
library(patchwork)
library(DT)         # For pretty html rendering of big tables.
library(scuttle)    # Usefule SCE functions liek 'aggregateAcrossCells'
library(scater)     # Usfele SCE functions like 'plotExpression'
library(limma)      # for differential expression
library(edgeR)      # for differential expression
library(spicyR)
```

## Paths

```{r}

spe_02_banksy_niches_file <- file.path("data/GSE234713_CosMx_IBD_sfe_mini_02_banksy_niches")

spe_pseudobulk_by_celltype_file           <- file.path("data/GSE234713_CosMx_IBD_sfe_mini_pseudobulk_by_celltype.RDS")
spe_pseudobulk_by_niche_file              <- file.path("data/GSE234713_CosMx_IBD_sfe_mini_pseudobulk_by_niche.RDS")
spe_pseudobulk_by_celltype_in_niche_file  <- file.path("data/GSE234713_CosMx_IBD_sfe_mini_pseudobulk_by_celltype_in_niche.RDS")

```



## Load data

Before we get started - load the preprocessed data
```{r}

# Load a spatialFeatureExperiment object (a type of SingleCellExperimet)
sfe <- readObject(spe_02_banksy_niches_file)

# And subset to one sample, for later plotting.
sfe.sample <- sfe[,sfe$tissue_sample == 'HC_b'] 
```





# The study

This workshop will use CosMx Spatial Molecular Imager (SMI) data from the paper "Macrophage and neutrophil heterogeneity at single-cell spatial resolution in human inflammatory bowel disease" (Garrido-Trigo et al. 2023) - link: https://www.nature.com/articles/s41467-023-40156-6

There were 9x colon tissue samples, one per slide. They used a 1k RNA panel panel. (5k, 6k and whole transcriptome available) 

* 3x Healthy controls (HC)
* 3x Crohnâ€™s disease (CD)
* 3x Ulcerative colitis (UC)

For this workshop, we will work with a subsetted dataset;

* Only the Healthy and Chrohn's disease samples - 6x samples total
* Only a regional subset of each sample - First 4 'FOVs' (feilds of view - cosmx scans)




This 'mini' dataset has 999 genes and 65601 cells. 

```{r}
sfe
```



And the following g

```{r}
DT::datatable(as.data.frame(colData(sfe)))
```











